File: apps/desktop/src/stores/skills.ts
import { create } from 'zustand';
import { invoke } from '@tauri-apps/api/core';

// ─── Types ──────────────────────────────────────────────────────────────────

export interface SkillSource {
  type: 'github' | 'local' | 'url';
  url: string;
  subpath?: string;
  ref?: string;
  provider?: string;
}

export interface AgentInstallRecord {
  agent: string;
  scope: 'global' | 'project';
  mode: 'symlink' | 'copy';
  installedPath: string;
  installedAt: string;
}

export interface Skill {
  id: string;
  name: string;
  description: string;
  source: SkillSource;
  canonicalPath: string;
  version: string;
  author: string;
  tags: string[];
  installs: AgentInstallRecord[];
  updatedAt: string;
  treeSha: string;
  contentHash: string;
}

export interface SkillDetail extends Skill {
  readmeContent: string;
  fileList: string[];
  sourceUrl: string;
}

export interface ResolvedSkill {
  name: string;
  description: string;
  source: SkillSource;
  version?: string;
  author?: string;
  tags?: string[];
}

export interface UpdateCheckResult {
  name: string;
  currentHash: string;
  latestHash: string;
  hasUpdate: boolean;
  currentVersion?: string;
  latestVersion?: string;
}

// ─── Store ──────────────────────────────────────────────────────────────────

interface SkillsState {
  // Data
  skills: Skill[];
  isLoading: boolean;
  error: string | null;

  // Filters
  selectedAgent: string;
  searchQuery: string;

  // Install modal
  installModalOpen: boolean;
  installSource: string;
  isInstalling: boolean;
  resolvedSkills: ResolvedSkill[];
  isResolving: boolean;
  resolveError: string | null;
  installError: string | null;

  // Update checks
  updateResults: UpdateCheckResult[];
  isCheckingUpdates: boolean;

  // Actions
  fetchSkills: () => Promise<void>;
  installSkill: (
    source: string,
    agents: string[],
    mode: string,
    scope: string,
    force?: boolean,
    agentBindings?: Record<string, unknown>
  ) => Promise<Skill | null>;
  removeSkill: (name: string, agents?: string[], cleanCache?: boolean) => Promise<void>;
  updateSkill: (name: string) => Promise<Skill | null>;
  checkUpdates: () => Promise<void>;
  resolveSource: (source: string) => Promise<void>;

  // Setters
  setSelectedAgent: (agent: string) => void;
  setSearchQuery: (query: string) => void;
  setInstallModalOpen: (open: boolean) => void;
  setInstallSource: (source: string) => void;
  clearResolvedSkills: () => void;
}

export const useSkillsStore = create<SkillsState>((set, get) => ({
  // Data
  skills: [],
  isLoading: false,
  error: null,

  // Filters
  selectedAgent: 'all',
  searchQuery: '',

  // Install modal
  installModalOpen: false,
  installSource: '',
  isInstalling: false,
  resolvedSkills: [],
  isResolving: false,
  resolveError: null,
  installError: null,

  // Update checks
  updateResults: [],
  isCheckingUpdates: false,

  // ─── Actions ────────────────────────────────────────────────────────

  fetchSkills: async () => {
    set({ isLoading: true, error: null });
    try {
      const { selectedAgent, searchQuery } = get();
      const params: Record<string, string> = {};
      if (selectedAgent !== 'all') params.agent = selectedAgent;
      // Removed scope filter - all skills are global now
      if (searchQuery.trim()) params.query = searchQuery.trim();

      const skills = await invoke<Skill[]>('list_skills', params);
      set({ skills, isLoading: false });
    } catch (err) {
      const errorMsg =
        err instanceof Error ? err.message : typeof err === 'string' ? err : JSON.stringify(err);
      set({ error: errorMsg, isLoading: false });
    }
  },

  installSkill: async (source, agents, mode, scope, force, agentBindings) => {
    set({ isInstalling: true, installError: null });
    try {
      const params: Record<string, unknown> = { source, agents, mode, scope };
      if (force !== undefined) params.force = force;
      if (agentBindings) params.agent_bindings = agentBindings;

      const skill = await invoke<Skill>('install_skill', params);
      // Refresh the list after install
      await get().fetchSkills();
      set({ isInstalling: false });
      return skill;
    } catch (err) {
      const errorMsg =
        err instanceof Error ? err.message : typeof err === 'string' ? err : JSON.stringify(err);
      set({ installError: errorMsg, isInstalling: false });
      return null;
    }
  },

  removeSkill: async (name, agents, cleanCache) => {
    set({ isLoading: true, error: null });
    try {
      const params: Record<string, unknown> = { name };
      if (agents) params.agents = agents;
      if (cleanCache !== undefined) params.clean_cache = cleanCache;

      await invoke('remove_skill', params);
      await get().fetchSkills();
    } catch (err) {
      const errorMsg =
        err instanceof Error ? err.message : typeof err === 'string' ? err : JSON.stringify(err);
      set({ error: errorMsg, isLoading: false });
    }
  },

  updateSkill: async (name) => {
    set({ isLoading: true, error: null });
    try {
      const skill = await invoke<Skill>('update_skill', { name });
      await get().fetchSkills();
      // Re-check updates to refresh the update badges
      await get().checkUpdates();
      return skill;
    } catch (err) {
      const errorMsg = String(err);
      console.error('Update skill error:', errorMsg);
      set({ error: errorMsg, isLoading: false });
      return null;
    }
  },

  checkUpdates: async () => {
    set({ isCheckingUpdates: true });
    try {
      const results = await invoke<UpdateCheckResult[]>('check_updates');
      set({ updateResults: results, isCheckingUpdates: false });
    } catch (err) {
      const errorMsg =
        err instanceof Error ? err.message : typeof err === 'string' ? err : JSON.stringify(err);
      set({ error: errorMsg, isCheckingUpdates: false });
    }
  },

  resolveSource: async (source) => {
    set({ isResolving: true, resolveError: null, resolvedSkills: [] });
    try {
      const resolved = await invoke<ResolvedSkill[]>('resolve_source', { source });
      set({ resolvedSkills: resolved, isResolving: false });
    } catch (err) {
      const errorMsg =
        err instanceof Error ? err.message : typeof err === 'string' ? err : JSON.stringify(err);
      set({ resolveError: errorMsg, isResolving: false });
    }
  },

  // ─── Setters ────────────────────────────────────────────────────────

  setSelectedAgent: (agent) => set({ selectedAgent: agent }),
  setSearchQuery: (query) => set({ searchQuery: query }),
  setInstallModalOpen: (open) =>
    set({
      installModalOpen: open,
      ...(open
        ? {}
        : {
            installSource: '',
            resolvedSkills: [],
            resolveError: null,
            installError: null,
            isResolving: false,
            isInstalling: false,
          }),
    }),
  setInstallSource: (source) => set({ installSource: source }),
  clearResolvedSkills: () => set({ resolvedSkills: [], resolveError: null, installError: null }),
}));

File: apps/desktop/src-tauri/src/commands/skills.rs
use crate::error::CommandError;
use crate::services::skill_import::{import_skills_from_agent, scan_and_import_skills, ScanResult};
use crate::services::skills::SkillEngine;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Skill {
    pub id: String,
    pub name: String,
    pub description: String,
    pub source: Source,
    #[serde(rename = "canonicalPath")]
    pub canonical_path: String,
    pub version: Option<String>,
    pub author: Option<String>,
    pub tags: Vec<String>,
    pub installs: Vec<AgentInstallRecord>,
    #[serde(rename = "updatedAt")]
    pub updated_at: String,
    #[serde(rename = "treeSha")]
    pub tree_sha: Option<String>,
    #[serde(rename = "contentHash")]
    pub content_hash: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Source {
    #[serde(rename = "type")]
    pub source_type: String,
    pub url: String,
    pub subpath: Option<String>,
    #[serde(rename = "ref")]
    pub git_ref: Option<String>,
    pub provider: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentInstallRecord {
    pub agent: String,
    pub scope: String,
    pub mode: String,
    #[serde(rename = "installedPath")]
    pub installed_path: String,
    #[serde(rename = "installedAt")]
    pub installed_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkillDetail {
    #[serde(flatten)]
    pub skill: Skill,
    #[serde(rename = "readmeContent")]
    pub readme_content: String,
    #[serde(rename = "fileList")]
    pub file_list: Vec<String>,
    #[serde(rename = "sourceUrl")]
    pub source_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResolvedSkill {
    pub name: String,
    pub description: Option<String>,
    pub source: Source,
    pub provider: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateCheckResult {
    pub name: String,
    #[serde(rename = "currentHash")]
    pub current_hash: String,
    #[serde(rename = "latestHash")]
    pub latest_hash: String,
    #[serde(rename = "hasUpdate")]
    pub has_update: bool,
    #[serde(rename = "currentVersion")]
    pub current_version: Option<String>,
    #[serde(rename = "latestVersion")]
    pub latest_version: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentBindingOverride {
    pub mode: Option<String>,
    pub scope: Option<String>,
}

#[tauri::command]
pub async fn list_skills(
    agent: Option<String>,
    scope: Option<String>,
    query: Option<String>,
) -> Result<Vec<Skill>, CommandError> {
    let engine = SkillEngine::new();
    Ok(engine.list_skills(agent.as_deref(), scope.as_deref(), query.as_deref()))
}

#[tauri::command]
pub async fn install_skill(
    source: String,
    agents: Vec<String>,
    mode: String,
    scope: String,
    force: Option<bool>,
    agent_bindings: Option<HashMap<String, AgentBindingOverride>>,
) -> Result<Skill, CommandError> {
    let engine = SkillEngine::new();
    engine
        .install_skill(
            &source,
            &agents,
            &mode,
            &scope,
            force.unwrap_or(false),
            agent_bindings.as_ref(),
        )
        .await
}

#[tauri::command]
pub async fn remove_skill(
    name: String,
    agents: Option<Vec<String>>,
    clean_cache: Option<bool>,
) -> Result<(), CommandError> {
    let engine = SkillEngine::new();
    engine
        .remove_skill(&name, agents.as_deref(), clean_cache.unwrap_or(false))
        .await
}

#[tauri::command]
pub async fn add_skill_to_agents(
    name: String,
    agents: Vec<String>,
    mode: Option<String>,
    scope: Option<String>,
) -> Result<Skill, CommandError> {
    let engine = SkillEngine::new();

    // Get the skill's source from lock file
    let skills = engine.list_skills(None, None, Some(&name));
    let skill = skills
        .into_iter()
        .find(|s| s.name == name)
        .ok_or_else(|| CommandError::SkillNotFound(name.clone()))?;

    // Install to the new agents using the existing source
    engine
        .install_skill(
            &skill.source.url,
            &agents,
            &mode.unwrap_or_else(|| "symlink".to_string()),
            &scope.unwrap_or_else(|| "global".to_string()),
            false, // Don't force reinstall
            None,  // No agent-specific overrides
        )
        .await
}

#[tauri::command]
pub async fn update_skill(name: String) -> Result<Skill, CommandError> {
    let engine = SkillEngine::new();
    engine.update_skill(&name).await
}

#[tauri::command]
pub async fn check_updates() -> Result<Vec<UpdateCheckResult>, CommandError> {
    let engine = SkillEngine::new();
    engine.check_updates().await
}

#[tauri::command]
pub async fn get_skill_detail(name: String) -> Result<SkillDetail, CommandError> {
    let engine = SkillEngine::new();
    engine.get_skill_detail(&name)
}

#[tauri::command]
pub async fn resolve_source(source: String) -> Result<Vec<ResolvedSkill>, CommandError> {
    let engine = SkillEngine::new();
    engine.resolve_source(&source).await
}

#[tauri::command]
pub async fn scan_skills() -> Result<ScanResult, CommandError> {
    scan_and_import_skills().await
}

#[tauri::command]
pub async fn import_skills(agent: String) -> Result<ScanResult, CommandError> {
    import_skills_from_agent(&agent).await
}
